rfc/rfc004-distributed-document-format.md

```diff
@@ -0,0 +1,95 @@
+cryptographic authentication and verification of its contents, guaranteed ordering and access control.
+
+### Status of document
+
+This document is currently a draft.
+
+### Copyright Notice
+
+![](../.gitbook/assets/license.png)
+
+This document is released under the [Attribution-ShareAlike 4.0 International \(CC BY-SA 4.0\) license](https://creativecommons.org/licenses/by-sa/4.0/).
+
+## 1.  Introduction
+
+Because Nuts is a decentralized network there needs to be a standardized format for structuring public information
+(like where to find another care organization's data) in a way that allows each party to independently verify
+the authenticity and integrity of that information. This document proposes such a format. Systems can use the format
+to build their application on top of it making it suitable for publishing on a distributed network.
+
+This document does not define how to transport data to other participants in a distributed network.
+
+## 2. Terminology
+
+* **Document**: piece of application data enveloped with metadata like signatures and references to other documents
+  published on a distributed network.
+
+Other terminology comes from the [Nuts Start Architecture](rfc001-nuts-start-architecture.md#nuts-start-architecture).
+
+## 3. Document Format
+Documents MUST be encoded as [RFC7515 JSON Web Signature](https://tools.ietf.org/html/rfc7515).
+
+### 3.1. JWS implementation
+In addition to required header parameters as specified in RFC7515 the following requirements apply:
+* **x5c**: MUST be present and contain exactly one entry with the X.509 signing certificate associated with the private key.
+  The certificate MUST have the digitalSignature and contentCommitment key usages and must be valid at time of signing.
+  The certificate SHOULD conform to RFC Certificate Structure (TODO) so other parties can validate it.
+* **alg**: MUST be one of the following algorithms: `PS256`, `PS384`, `PS512`, `ES256`, `ES384` or `ES512`.
+  other algorithms SHALL NOT be used.
+* **cty**: MUST contain the type of the payload indicating how to interpret the payload encoded as string.
+* **crit** MUST contain the **sigt** and **version** headers.
+
+The **jku**, **jwk**, **kid** and **x5u** header parameters SHOULD NOT be used and MUST be ignored by when processing the document.
+
+In addition to the registered header parameters, the following header MUST be present as protected headers:
+* **sigt**: (signing time) MUST contain the signing time of the document in UTC as string, formatted according to [RFC3339](https://tools.ietf.org/html/rfc3339).
+* **version**: MUST contain the format version of the document as number. For this version of the format the version MUST be 1.
+
+The following headers are OPTIONAL but must be a protected header when present:
+* **prev**: MUST (if present) contain the reference (see section 3.2) of the preceding document.
```

Comment:
What if nodes just ignore genesis documents (without a prev) when they've already got one? Then you *have* to have another vendor register your vendor CA certificate (as planned) and you can't publish data without having pulled the latest network state (last hash) from the network since your genesis document will get ignored anyways? And when joining, the safest bet is probably to connect to your friend's vendor who published your vendor CA certificate since you think you can trust him/her to provide you with the correct chain of documents.

Just leaves the problem; what to do with branches and ordering. Generally not a problem since the data vendors register don't interfere with each other unless someone tries to register your vendor earlier in the document chain. Let's say you registered your vendor on document #10 (D10) and someone else claims your vendor with stolen proof on D5. Maybe some kind of merging like 'this was the previous doc but I also acknowledge this branch'. Makes things more complicated though.

Reply:
Added part of our discussion to this section (genesis doc), vendor CA trust will end up in RFC006.